src/ng/parse.js

```diff
@@ -622,15 +622,51 @@ function isStateless($filter, filterName) {
   return !fn.$stateful;
 }
 
-function findConstantAndWatchExpressions(ast, $filter) {
+// Detect nodes which could depend on non-shallow state of objects
+function isShallowState(stack) {
+  for (var i = 0; i < stack.length; i++) {
```

Comment:
Yeah that's probably true.  Probably store the result of the switch statement in the stack? I'll try it out...